File: problem/Lexicographical_String/getHappyString.js
/**
 * @param {number} n
 * @param {number} k
 * @return {string}
 */
var getHappyString = function(n, k) {

    const total = 3 * (1 << (n-1))
    if(k > total) return ""

    k--

    let res=""
    let last=""

    for(let pos=0; pos<n; pos++){

        const branch = 1 << (n-pos-1)

        const choices = ['a','b','c'].filter(c => c!==last)

        const idx = Math.floor(k / branch)

        res += choices[idx]
        last = choices[idx]

        k %= branch
    }

    return res
};

File: problem/Lexicographical_String/getHappyString.c
char *getHappyString(int n, int k)
{
    int total = 3 * (1 << (n - 1));
    if (k > total)
        return "";

    char *res = (char *)malloc((n + 1) * sizeof(char));
    res[n] = '\0';

    k--;
    char last = '\0';

    for (int pos = 0; pos < n; pos++)
    {
        int branch = 1 << (n - pos - 1);

        char choices[3];
        int count = 0;
        for (char c = 'a'; c <= 'c'; c++)
        {
            if (c != last)
            {
                choices[count++] = c;
            }
        }

        int idx = k / branch;
        res[pos] = choices[idx];
        last = choices[idx];
        k %= branch;
    }

    return res;
}